### PR TITLE
[7.x] add `everyXMinutes()` and `everyXHours()`

### PR DESCRIPTION
**Pull request by browner12**
Recently a large number of methods have been added for specific minute and hour scheduling intervals:

```php
->everyTwoMinutes();
->everyThreeMinutes();
->everyTwoHours();
->everyThreeHours();
etc...
```

This PR creates more generic methods that allow passing a parameter:

```php
->everyXMinutes(2);
->everyXMinutes(3);
->everyXHours(2);
->everyXHours(3);
```

This keeps the LOCs down and maintenance a little easier. 

It also makes available the intervals that are currently missing from the new methods:

```php
->everyXMinutes(6);
->everyXMinutes(12);
->everyXMinutes(20);
->everyXHours(8);
->everyXHours(12);
```

I previously submitted this [PR](33380).  Taylor's concern there was that numbers that were not factors of 60 would cause unexpected results.  This PR solves that by requiring the parameter to be a factor of either 60 (for minutes) or 24 (for hours), and throws an `InvalidArgumentException()` otherwise.

### CODE REVIEW

File: src/Illuminate/Console/Scheduling/ManagesFrequencies.php [hunk 1]
```diff
@@ -71,6 +71,21 @@ private function inTimeInterval($startTime, $endTime)
         };
     }
 
+    /**
+     * Schedule the event to run every minute that is divisible by the interval.
+     *
+     * @param  int  $minutes
+     * @return $this
+     */
```
Comment (GrahamCampbell) [inline src/Illuminate/Console/Scheduling/ManagesFrequencies.php:79]: missing throws doc

File: src/Illuminate/Console/Scheduling/ManagesFrequencies.php [hunk 2]
```diff
@@ -151,6 +166,22 @@ public function everyThirtyMinutes()
         return $this->spliceIntoPosition(1, '0,30');
     }
 
+    /**
+     * Schedule the event to run every hour that is divisible by the interval.
+     *
+     * @param  int  $hours
+     * @return $this
+     */
```
Comment (GrahamCampbell) [inline src/Illuminate/Console/Scheduling/ManagesFrequencies.php:174]: same

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Comment by GrahamCampbell** [comment]
Would anyone actually use this method. I'd say don't add it until someone actually wants it.

**Comment by browner12** [comment]
Somebody is already using it, with the newly merged `->everyTwoMinutes()` etc methods.  This is just a proposed way that is less code, and offers the full gamut of available intervals.

**Comment by SjorsO** [comment]
Something to consider is that you don't typically run/test the `schedule()` method of the console kernel during development. This means you won't notice that `->everyXHours(7)` throws an exception until after you've deployed your code.

**Comment by GrahamCampbell** [comment]
> Somebody is already using it,

Well, no. They are using the methods that were already added. Who is going to use your new methods?

**Comment by GrahamCampbell** [comment]
👎 from me. Moreover, the point about people not knowing if their X value is right is a concern too, Better to hard code every possible value as a method name than to have a dynamic method, IMO. Not that we will do that, either.

TLDR: We don't need to add these methods, or any others. The current state of affairs is already fine.

**Comment by jmarcher** [comment]
We already have `->cron('* * * * *')` if fine-tuning is necessary

**Comment by lk77** [comment]
Hello,

from my point of view, those methods needs to be added, and the old methods (everyTwoMinutes, everyThreeHours...) needs to be deprecated and removed in 8.x. It's a lot more logical to have minutes / hours in a parameter and not in the function name.

**Comment by taylorotwell** [comment]
I disagree the everyTwoMinutes style methods need to be removed. As we can see from the tests there are actually not that many possible values. I would honestly rather just have methods for every possible factor rather than generic methods. 🤷  Maybe I'm weird.

**Comment by rcerljenko** [comment]
@GrahamCampbell this is a terrible argument:
```
Moreover, the point about people not knowing if their X value is right is a concern too, Better to hard code every possible value as a method name than to have a dynamic method, IMO. Not that we will do that, either.
```
Just beacause you don't uderstand the importance of generalized methods doesn't mean that is a bad idea. In my humble opinion hardcoding "one", "two", "three" (and more) methods is a poor pattern and should be avoided.
@lk77 has a point here (together with the creator of this PR)

**Comment by GrahamCampbell** [comment]
> this is a terrible argument:

No it's not. It's an argument about correctness of the code. Hard coded methods are something people know statically will work. With these x methods, you only find out your code is broken at runtime (in production).

**Comment by GrahamCampbell** [comment]
The generalised method would be fine if all input values were fine, but they're not...

**Comment by SjorsO** [comment]
I would open another PR to add methods for all other missing values, but I think I've already ruffled enough feathers with my first two PRs :smile:

**Comment by GrahamCampbell** [comment]
:trollface:

**Comment by browner12** [comment]
I mean, all the inputs are technically fine, as long as people understand how they work.

`->everyXMinutes(17)` still works, just not how you would intuitively expect, so you gotta read the docs.

At this point, I say just round them out with all the methods if Taylor doesn't want this generic one, so then we don't have to worry about it anymore.

Minutes: 1, 2, 3, 4, 5, 6, 10, 12, 15, 20, 30
Hours: 1, 2, 3, 4, 6, 8, 12

**Comment by lk77** [comment]
Well, if somebody want's 17 minutes, he will use cron('*/17 * * * *') which will execute :

```
2020-07-03 09:00:00
2020-07-03 09:17:00
2020-07-03 09:34:00
2020-07-03 09:51:00
2020-07-03 10:00:00
....
```

it works the same, but it's a lot less straightforward,
and it repeat each hour with only 9 minutes between the last execution within one hour and the first execution of the next hour.
There is no solution if the user wants an awkward value
